tpm2tools/keys.go

```diff
@@ -240,39 +272,45 @@ func (k *Key) Unseal(in *proto.SealedBytes) ([]byte, error) {
 	}
 	defer tpm2.FlushContext(k.rw, sealed)
 
-	if len(sel.PCRs) == 0 {
-		return tpm2.Unseal(k.rw, sealed, "")
+	if cOpt != nil {
+		var ticket tpm2.Ticket
+		_, err = tpmutil.Unpack(in.GetTicket(), &ticket)
+		if err != nil {
+			return nil, fmt.Errorf("Ticket unpack failed: %v", err)
+		}
+		hashCon := sessionHashAlg.New()
+		hashCon.Write(in.GetCreationData())
+		creationDataHash := hashCon.Sum(nil)
+		_, _, err = tpm2.CertifyCreation(k.rw, "", sealed, tpm2.HandleNull, nil, creationDataHash, tpm2.SigScheme{}, ticket)
+		if err != nil {
+			return nil, fmt.Errorf("failed to certify creation: %v", err)
+		}
+		err = cOpt.CertifyPCRs(k.rw, in.GetCertifiedPcrs())
+		if err != nil {
+			return nil, fmt.Errorf("failed to certify PCRs: %v", err)
+		}
 	}
 
+	sel := tpm2.PCRSelection{Hash: tpm2.Algorithm(in.Hash)}
+	for _, pcr := range in.Pcrs {
+		sel.PCRs = append(sel.PCRs, int(pcr))
+	}
 	session, err := createPCRSession(k.rw, sel)
 	if err != nil {
-		return nil, fmt.Errorf("failed to unseal: %v", err)
+		return nil, fmt.Errorf("failed to create session: %v", err)
 	}
 	defer tpm2.FlushContext(k.rw, session)
 
 	return tpm2.UnsealWithSession(k.rw, session, sealed, "")
 }
 
-// Reseal unwraps a secret and rewraps it under the auth value that would be
-// produced by the PCR state in pcrs. Similar to seal and unseal, this acts on
-// the SHA256 PCRs and uses the owner hierarchy.
+// Reseal unwraps a secret using cOpt and then reseal the secret using sOpt
+// cOpt and sOpt can be nil
```

Comment:
This description can now be shorter, just say it's a helper to do `Unseal()` then `Seal()`. Keep most of the detail in the functions above.

Reply:
// Reseal is a shortcut to call Unseal() followed by Seal().
// CertificationOpt will be used in Unseal(), and SealingOpt wil be used in Seal()